nautobot/core/apps/__init__.py

```diff
@@ -0,0 +1,249 @@
+
+
+registry["nav_menu"] = {"tabs": {}}
+
+
+class NautobotConfig(AppConfig):
+    """
+    Custom AppConfig for Nautobot application.
+
+    Adds functionality to generate the HTML navigation menu using `navigation.py` files from nautbot
+    applications.
+    """
+
+    menu_tabs = "navigation.menu_tabs"
+
+    def ready(self):
+        """
+        Ready function initiates the import application.
+        """
+        menu_items = import_object(f"{self.name}.{self.menu_tabs}")
+        if menu_items is not None:
+            register_menu_items(menu_items)
+
+
+def register_menu_items(tab_list):
+    """
+    Using the imported object a dictionary is either created or updated with objects to create
+    the navbar.
+
+    The dictionary is built from four key objects, NavMenuTab, NavMenuGroup, NavMenuItem and
+    NavMenuButton. The Django template then uses this dictionary to generate the navbar HTML.
+    """
+    for nav_tab in tab_list:
+        if nav_tab.name not in registry["nav_menu"]["tabs"]:
+            registry["nav_menu"]["tabs"][nav_tab.name] = {
+                "weight": nav_tab.weight,
+                "groups": {},
+                "permissions": [],
+            }
+
+        tab_perms = []
+        registry_groups = registry["nav_menu"]["tabs"][nav_tab.name]["groups"]
+        for group in nav_tab.groups:
+            if group.name not in registry_groups:
+                registry_groups[group.name] = {
+                    "weight": group.weight,
+                    "items": {},
+                    "permissions": [],
+                }
+
+            group_perms = []
+            for item in group.items:
+                if item.link not in registry_groups[group.name]["items"]:
+                    registry_groups[group.name]["items"][item.link] = {
+                        "link_text": item.link_text,
+                        "weight": item.weight,
+                        "buttons": {},
+                        "permissions": item.permissions,
+                    }
+
+                    registry_buttons = registry_groups[group.name]["items"][item.link]["buttons"]
+                    for button in item.buttons:
+                        if button.title not in registry_buttons:
+                            registry_buttons[button.title] = {
+                                "button_class": button.button_class,
+                                "icon_class": button.icon_class,
+                                "link": button.link,
+                                "permissions": button.permissions,
+                                "weight": button.weight,
+                            }
+
+                    # Add sorted buttons to group registry dict
+                    registry_groups[group.name]["items"][item.link]["buttons"] = OrderedDict(
+                        sorted(registry_buttons.items(), key=lambda kv_pair: kv_pair[1]["weight"])
+                    )
```

Comment:
I think this whole block needs one less level of indentation - it should happen regardless of the `item.link not in registry_groups...` check, I believe?